### PR TITLE
Fix AdaptiveThresholdProcessor throws IndexOutOfRangeException

### PR DESCRIPTION
**Pull request by brianpopow**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

This PR fixes issue #2217.

The PR is still marked as WIP, because I am not sure the output is correct. It's all black. I assume that's not how it should look like? I am not familiar with the algorithm.

Input: 
187073251-5d23775b-f4b9-459b-a686-e2556401a1f7

Output:
output

### CODE REVIEW

File: src/ImageSharp/Processing/Processors/Binarization/AdaptiveThresholdProcessor{TPixel}.cs [hunk 1]
```diff
@@ -133,19 +133,21 @@ public void Invoke(int y)
...
                 for (int x = this.startX; x < this.endX; x++)
                 {
                     TPixel pixel = pixelRow[x];
                     pixel.ToRgba32(ref rgb);
 
-                    var x1 = Math.Max(x - this.startX - this.clusterSize + 1, 0);
-                    var x2 = Math.Min(x - this.startX + this.clusterSize + 1, this.bounds.Width - 1);
-                    var y1 = Math.Max(y - this.startY - this.clusterSize + 1, 0);
-                    var y2 = Math.Min(y - this.startY + this.clusterSize + 1, this.bounds.Height - 1);
+                    int x1 = Math.Min(Math.Max(x - this.startX - this.clusterSize + 1, 0), maxX);
```
Comment (gfoidl) [inline src/ImageSharp/Processing/Processors/Binarization/AdaptiveThresholdProcessor{TPixel}.cs:144]: Nit: could use `Math.Clamp`?

**Comment by JimBobSquarePants** [comment]
>I am not familiar with the algorithm.

Unfortunately me neither. It was added [here](https://github.com/SixLabors/ImageSharp/pull/725). I'll read the attached [paper](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.420.7883&rep=rep1&type=pdf). 

One thing I have noticed having a quick look at the code is that it does not use our standardized Integral Image extension method added [here](https://github.com/SixLabors/ImageSharp/pull/1561). I'll pull it down and have a fiddle.

**Comment by JimBobSquarePants** [comment]
Looks like I have this working.

AdaptiveThreshold_WithRectangle_Works_Rgba32_Bradley02
AdaptiveThreshold_Works_Rgba32_Bradley01
AdaptiveThreshold_Works_Rgba32_Bradley02
AdaptiveThreshold_Works_Rgba32_ducky
Issue_2217_AdaptiveThreshold_DoesNotThrowIndexOutOfRangeException_Rgba32_Issue_2217_AdaptiveThresholdProcessor

**Comment by brianpopow** [comment]
> Looks like I have this working.
> 

Nice, good job in solving this issue! I think this is now ready to be merged.

**Pull request merged**